### PR TITLE
Parser refactor

### PR DESCRIPTION
**Pull request by CharlesChen0823**

### CODE REVIEW

File: src/html5/parser/helper.rs [hunk 1]
```diff
@@ -0,0 +1,501 @@
...
+        override_node: Option<NodeId>,
+    ) -> InsertionPositionMode<NodeId> {
+        let current_node_id = current_node!(self).id;
+        let target_id = override_node.unwrap_or(current_node_id);
+        let target_node = get_node_by_id!(self.document, target_id);
+        if !(self.foster_parenting
+            && ["table", "tbody", "thead", "tfoot", "tr"].contains(&target_node.name.as_str()))
+        {
+            if target_node.name == "template" {
+                panic!("current not support");
```
Comment (CharlesChen0823) [inline src/html5/parser/helper.rs:264]: current remove template content implemented.

**Comment by jaytaph** [comment]
@CharlesChen0823 When you are ready with this one, I'll approve and merge it so we can merge my PR on top of that.

**Comment by CharlesChen0823** [comment]
@jaytaph I had add template case in appropriate place for insert, but I think that part need more thinking or you can help me.  
Currently, I modify the state of pr for ready merge.

**Review by jaytaph — APPROVED** [review]

**Comment by jaytaph** [comment]
@CharlesChen0823 Looks good! I see there is some overlap with my PR, but i'll rebase it on top of this.

**Pull request merged**